main.go

```diff
@@ -140,6 +176,90 @@ func main() {
 	}
 }
 
+func deleteAction(ctx context.Context, cancel context.CancelFunc, clientSet client.Client, hr *fluxhelmv2beta1.HelmRelease, timeout, interval time.Duration) {
+	defer cancel()
+	instance := &fluxhelmv2beta1.HelmRelease{}
+	key := client.ObjectKey{
+		Name:      hr.Name,
+		Namespace: hr.Namespace,
+	}
+	if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+		log.Errorf("Failed to get instance of the helmrelease with %v", err)
+		return
+	} else if err != nil {
+		// Unexpectedly, the object that we are trying to delete was not found
+		// If this happens, we will log an error and return
+		log.Errorf("Did not find the helm release: %v", hr.Name)
+		return
+	} else {
+		log.Infof("Found the helm release: %v, deleting the release...", hr.Name)
+		// Delete the HelmRelease
+		if err := retry(ctx, func() error { return clientSet.Delete(ctx, hr) }, interval); err != nil {
+			log.Errorf("retry got err: %v", err)
+			return
+		}
+	}
+
+	pollingFunc := func() bool {
+		instance := &fluxhelmv2beta1.HelmRelease{}
+		key := client.ObjectKey{
+			Name:      hr.Name,
+			Namespace: hr.Namespace,
+		}
+		if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+			log.Errorf("Failed to get instance of the helmrelease with %v", err)
+			return false
+		} else if err != nil {
+			log.Infof("Did not find the helm release: %v, the object is deleted", hr.Name)
+			return true
+		} else {
+			log.Infof("Found the helm release: %v, the object is not deleted", hr.Name)
+			return false
+		}
+	}
+
+	if err := poll(ctx, pollingFunc, time.Second); err != nil {
+		log.Errorf("Failed to poll the deletion of the helm release with %v", err)
+		// The helm release is still not deleted. Force cleanup as the final attempt
+		forceCleanupHelmRelease(clientSet, hr, interval)
+	}
+}
+
+// remove the finalizer from the helm release and execute force uninstall
+func forceCleanupHelmRelease(clientSet client.Client, hr *fluxhelmv2beta1.HelmRelease, interval time.Duration) {
+	// Create a new context
+	ctx := context.Background()
+	instance := &fluxhelmv2beta1.HelmRelease{}
+	key := client.ObjectKey{
+		Name:      hr.Name,
+		Namespace: hr.Namespace,
+	}
+	if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+		log.Errorf("Failed to get instance of the helmrelease with %v", err)
+	} else if err != nil {
+		// The helm release was not found after the first failed cleanup attempt
+		log.Errorf("Did not find the helm release: %v, the object is deleted", hr.Name)
+	} else {
+		log.Infof("Removing the helm release finalizer: %v...", fluxhelmv2beta1.HelmReleaseFinalizer)
+		controllerutil.RemoveFinalizer(instance, fluxhelmv2beta1.HelmReleaseFinalizer)
+		if err := retry(ctx, func() error { return clientSet.Update(ctx, instance) }, interval); err != nil {
```

Comment:
@davutkarabay you're right - this will never timeout. Can we do `Patch()` instead of the `Update()` once and let it move on.

Reply:
Patch() instead of Update() makes sense to me - so I changed it to Patch(). Though I am not sure why update needs retry, but not patch? Is it because a simple patch like this is probably much more efficient than update, and is therefore much less likely to fail?

Reply:
If I am not wrong, patch uses a merge strategy to change the object while update does a blind blanket replace